File: src/main/java/dev/sky_lock/menu/MenuContents.java
package dev.sky_lock.menu;

import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author sky_lock
 */

public abstract class MenuContents {
    private final List<Slot> slots = new ArrayList<>();

    public void close() {
        slots.clear();
    }

    public void addSlot(Slot... slots) {
        Collections.addAll(this.slots, slots);
    }

    public void updateItemStack(int index, ItemStack newStack) {
        slots.stream().filter(slot -> slot.index() == index).findFirst().ifPresent(slot -> slot.setItemStack(newStack));
    }

    public List<Slot> getSlots() {
        return slots;
    }

    public int size() {
        return slots.size();
    }

    public void clear() {
        slots.clear();
    }

    void click(InventoryClickEvent event) {
        slots.stream().filter(slot -> slot.index() == event.getSlot()).findFirst().ifPresent(slot -> slot.click(event));
    }

    public abstract void onFlip(InventoryMenu menu);


}

File: src/main/kotlin/dev/sky_lock/menu/MenuContents.kt
package dev.sky_lock.menu

import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import java.util.*

/**
 * @author sky_lock
 */
abstract class MenuContents {
    private val slots: MutableList<Slot> = ArrayList()
    fun close() {
        slots.clear()
    }

    fun addSlot(vararg slots: Slot) {
        Collections.addAll(this.slots, *slots)
    }

    fun updateItemStack(index: Int, newStack: ItemStack) {
        slots.stream().filter { slot: Slot -> slot.index() == index }.findFirst().ifPresent { slot: Slot -> slot.setItemStack(newStack) }
    }

    fun getSlots(): List<Slot> {
        return slots
    }

    fun size(): Int {
        return slots.size
    }

    fun removeSlot(index: Int) {
        slots.stream().filter { slot: Slot -> slot.index() == index }.findFirst().ifPresent { o: Slot -> slots.remove(o) }
    }

    fun click(event: InventoryClickEvent) {
        slots.stream().filter { slot: Slot -> slot.index() == event.slot }.findFirst().ifPresent { slot: Slot -> slot.click(event) }
    }

    abstract fun onFlip(menu: InventoryMenu)
}